config/opal_setup_cc.m4

```diff
@@ -47,30 +41,42 @@ AC_DEFUN([OPAL_PROG_CC_C11_HELPER],[
-                   [Whether C compiler support atomic convenience variables in stdatomic.h])
+                   [[#include <stdatomic.h>]], [[static atomic_long foo = 1;++foo;]])
 
     OPAL_CC_HELPER([if $CC $1 supports C11 _Atomic keyword], [opal_prog_cc_c11_helper__Atomic_available],
-                   [[#include <stdatomic.h>]],[[static _Atomic long foo = 1;++foo;]], [OPAL_C_HAVE__ATOMIC],
-                   [Whether C compiler supports __Atomic keyword])
+                   [[#include <stdatomic.h>]],[[static _Atomic long foo = 1;++foo;]])
 
     OPAL_CC_HELPER([if $CC $1 supports C11 _Generic keyword], [opal_prog_cc_c11_helper__Generic_available],
-                   [[#define FOO(x) (_Generic (x, int: 1))]], [[static int x, y; y = FOO(x);]], [OPAL_C_HAVE__GENERIC],
-                   [Whether C compiler supports __Generic keyword])
+                   [[#define FOO(x) (_Generic (x, int: 1))]], [[static int x, y; y = FOO(x);]])
 
     OPAL_CC_HELPER([if $CC $1 supports C11 _Static_assert], [opal_prog_cc_c11_helper__static_assert_available],
-                   [[#include <stdint.h>]],[[_Static_assert(sizeof(int64_t) == 8, "WTH");]], [OPAL_C_HAVE__STATIC_ASSERT],
-                   [Whether C compiler support _Static_assert keyword])
+                   [[#include <stdint.h>]],[[_Static_assert(sizeof(int64_t) == 8, "WTH");]])
 
     dnl At this time Open MPI only needs thread local and the atomic convenience types for C11 support. These
     dnl will likely be required in the future.
-    AS_IF([test "x$opal_prog_cc_c11_helper__Thread_local_available" = "xyes" && test "x$opal_prog_cc_c11_helper_atomic_var_available" = "xyes"],
+    AS_IF([test $opal_prog_cc_c11_helper__Thread_local_available -eq 1 && test $opal_prog_cc_c11_helper_atomic_var_available eq 1],
           [$2], [$3])
 
+    AS_IF([test "$4" = "yes" ||
+          (test $opal_prog_cc_c11_helper__Thread_local_available -eq 1 && test $opal_prog_cc_c11_helper_atomic_var_available eq 1)],
+          [AC_DEFINE_UNQUOTED([OPAL_C_HAVE__THREAD_LOCAL], [$opal_prog_cc_c11_helper__Thread_local_available],
+                              [Whether C compiler supports __Thread_local])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE_ATOMIC_CONV_VAR], [$opal_prog_cc_c11_helper_atomic_var_available],
+                              [Whether C compiler support atomic convenience variables in stdatomic.h])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__ATOMIC], [$opal_prog_cc_c11_helper__Atomic_available],
+                              [Whether C compiler supports __Atomic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__GENERIC], [$opal_prog_cc_c11_helper__Generic_available],
+                              [Whether C compiler supports __Generic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__STATIC_ASSERT], [$opal_prog_cc_c11_helper__static_assert_available],
+                              [Whether C compiler support _Static_assert keyword])])
+
```

Comment:
I'm not sure I understand this block -- don't we *always* want to AC_DEFINE this values (and make sure that they are 0 or 1)?

Reply:
Hmm, indeed. We do. I haven't tested this with a non-C11 compiler. @ggouaillardet Might be worth doing. Anything older than gcc 4.9.0 will work. I have 4.8.x installed specifically to test with.

Reply:
i run a CentOS 7 box with gcc 4.8.x, which is a non-C11 compiler.

In this environment and on `master`, `OPAL_PROG_CC_C11_HELPER` is called several times in `OPAL_PROG_CC_C11` (test the C compiler without flag and then with all flags since all these test fail). Finally, `OPAL_PROG_CC_C11_HELPER` is invoked a last time in `OPAL_SETUP_CC` since `$opal_cv_c11_supported = no`.

With a C11 compiler, then `OPAL_PROG_CC_C11_HELPER` is invoked until it success.

I did not change this behavior in this PR.
In order to __always__ `AC_DEFINE` the macros, I added a forth argument to force the macros to be `AC_DEFINE`'ed, and this argument is only `yes` in `OPAL_SETUP_CC` since we know at this stage we do not have a C11 compiler.

Bottom line, I think this PR is correct.
And that being said, we might want to take this opportunity to revisit the existing logic.
Feel free to suggest anything and I will be happy to update this PR.